### PR TITLE
[ci] docker builds via buildkite

### PR DESCRIPTION
**Pull request by schallert**
Up until now we've been using Quay automated builds, which doesn't give
us all the granularity we want in tags. That is, every build Quay
creates it tags `latest` and `master` depending on the ref it builds
from.

What we really want is `master` to always point to `origin/master` and
`latest` to point to the latest tagged release, along with having a tag
for that release.

This PR modifies our buildkite pipeline to check if `HEAD` is
`origin/master` OR tagged, and if so dynamically add another step to the
pipeline to actually do the docker build. This minimizes blocking in the
queue for other jobs.

### CODE REVIEW

**Comment by schallert** [comment]
Had to close and reopen this PR to get CI working (FYI @robskillington )

File: .buildkite/scripts/check_do_docker.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+set -exuo pipefail
+
+# This script checks if we should do a docker build, and if so adds a step in
```
Comment (prateek) [inline .buildkite/scripts/check_do_docker.sh]: nit: mind adding some kind of check to ensure this is only attempted in on a CI host - something like `[[ -z BUILDKITE ]] || (echo "..."; exit 1)`

File: docker/images.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (prateek) [inline docker/images.json]: mind adding a readme in this directory explaining what/where this file is used

Comment (prateek) [reply docker/images.json]: nvm just saw the other files in the diff

File: .buildkite/docker-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+steps:
```
Comment (prateek) [inline .buildkite/docker-pipeline.yml]: nit: mind renaming file to `image-release-pipeline.yml` or something else to indicate what it's for

File: docker/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+3. For all images, and for all releases, there will be an image tagged `${IMAGE}:${RELEASE}`.
+
+## Builds
+
+This directory contains the Dockerfiles, configs, and build scripts for building images according to the above policy. `images.json` contains the base repo that will be used to tag images, as well as the config for each image. For example, with the following config:
+
+```json
+{
+  "image_base": "quay.io/m3db",
```
Comment (prateek) [inline docker/README.md]: nit: mind using `quay.io/m3` as the example in this file

File: docker/build.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  if [[ -z "$DRYRUN" ]]; then
+    docker push "$1"
+  else
+    echo "would push $1"
+  fi
+}
+
+CONFIG="docker/images.json"
+
+IMAGES="$(<$CONFIG jq -er '.images | to_entries | map(.key)[]')"
```
Comment (prateek) [inline docker/build.sh]: could you ensure `[[ -f $CONFIG ]]` before this

File: docker/build.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# If the current commit is exactly origin/master, push a tag for "master".
+if [[ "$CURRENT_SHA" == "$MASTER_SHA" ]]; then
+  TAGS_TO_PUSH="${TAGS_TO_PUSH} master"
+fi
+
+if [[ -z "$TAGS_TO_PUSH" ]]; then
+  exit 0
+fi
+
+for IMAGE in $IMAGES; do
```
Comment (prateek) [inline docker/build.sh]: nit: could you add a little more logging to the loops below

File: docker/build.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+set -exo pipefail
+
+function cleanup() {
+  docker system prune -f
+  find /tmp -name '*m3-docker' -print0 | xargs -0 rm -fv
+}
+
+trap cleanup EXIT
+
+function push() {
```
Comment (prateek) [inline docker/build.sh]: nit: mind calling this pushimage or something longer than just push. The usage of this in the loop made me think it was related to `pushd` until I remembered this decl

**Pull request merged**